src/harvest.cpp

```diff
@@ -89,7 +113,59 @@ void harvest_list::check_consistency()
         if( !errors.empty() ) {
             debugmsg( "Harvest list %s has invalid drop(s): %s", hl.id_.c_str(), errors.c_str() );
         }
+
+        for( const auto &e : hl.reqs ) {
+            if( !( e.first.is_null() || e.first.is_valid() ) || e.second < 0 ) {
+                debugmsg( "harvest list %s has unknown or incorrectly specified requirements %s",
+                          hl.id_.c_str(), e.first.c_str() );
+            }
+        }
+    }
+}
+
+requirement_data harvest_list::requirements() const
+{
+    return std::accumulate( reqs.begin(), reqs.end(), requirement_data(),
+        []( const requirement_data &lhs, const std::pair<requirement_id, int> &rhs ) {
+        return lhs + ( *rhs.first * rhs.second );
+    } );
+}
+
+bool harvest_list::can_harvest( const inventory &inv, bool alert ) const {
+    const requirement_data r = requirements();
+
+    for( const auto &opts : r.get_qualities() ) {
+        for( const auto &qual : opts ) {
+            if( !qual.has( inv ) ) {
+                if( alert ) {
+                    add_msg( m_info, _( "To perform this action you need %s" ), qual.to_string().c_str() );
+                }
+                return false;
+            }
+        }
     }
+
+    for( const auto &opts : r.get_tools() ) {
+        if( !std::any_of( opts.begin(), opts.end(), [&]( const tool_comp & tool ) {
+        return ( tool.count <= 0 && inv.has_tools( tool.type, 1 ) ) ||
+               ( tool.count  > 0 && inv.has_charges( tool.type, tool.count ) );
+        } ) ) {
```

Comment:
1. Introduce `tool_comp::has_amount_in( const inventory &inv ) const` and move the internals there.
2. Call the new function from both here and `player::can_disassemble()`.
3. Assign the lambda to a temp. variable so that the condition can become a single-liner.
4. Optionally (probably later) rename `tool_comp::has()` to `has_quality_in()` (for consistency).